### ISSUE TITLE
unable to load data exported by dump

### ISSUE THREAD

**Issue by zsalab** [cockroachdb/cockroach#51969]
**Describe the problem**

Put long rows into a table cause the cocokroach sql does not able to load the file created by cockroach dump

**To Reproduce**

```
# BS=1 # small rows
BS=1024 # large rows

rm -rf /tmp/cockroach-dump-test
mkdir -p /tmp/cockroach-dump-test/store /tmp/cockroach-dump-test/dump
echo '---- START ----'
cockroach start-single-node --insecure --listen-addr=localhost:26257 --store=/tmp/cockroach-dump-test/store --background

echo 'DROP DATABASE IF EXISTS dumptest CASCADE;' > /tmp/cockroach-dump-test/prepare-test.sql
echo 'CREATE DATABASE dumptest;' >> /tmp/cockroach-dump-test/prepare-test.sql
echo 'USE dumptest;' >> /tmp/cockroach-dump-test/prepare-test.sql
echo 'CREATE TABLE dumptest (id INT NOT NULL PRIMARY KEY, data VARCHAR);' >> /tmp/cockroach-dump-test/prepare-test.sql
for i in $(seq 1 10); do echo "INSERT INTO dumptest VALUES('$i', '$(dd if=/dev/urandom bs=$BS count=8192 | base64)');" >> /tmp/cockroach-dump-test/prepare-test.sql; done
                                                                    
echo '---- PREPARE ----'                                            
cockroach sql --insecure --host=localhost:26257 < /tmp/cockroach-dump-test/prepare-test.sql
                                                                    
echo 'USE dumptest;' > /tmp/cockroach-dump-test/dump.sql            
echo 'TRUNCATE TABLE dumptest;' >> /tmp/cockroach-dump-test/dump.sql        
echo '---- DUMP ----'                                                       
cockroach dump --insecure --host=localhost:26257 dumptest --dump-mode data >> /tmp/cockroach-dump-test/dump.sql
                                                                            
echo '---- LOAD ----'                                                       
cockroach sql --insecure --host=localhost:26257 < /tmp/cockroach-dump-test/dump.sql
                                                                            
echo '---- STOP ----'                                                       
cockroach quit --insecure --host=localhost:26257                            
```

**Expected behavior**

In every case what is exported by cockroach dump should be able to load back to the same cluster (or another one running with the same settings where the dump created)

**Additional data / screenshots**

loading the file cause this error (if use BS=1 in the script above no error appear): 

SET
Time: 1.007ms
TRUNCATE
Time: 529.081ms
ERROR: write tcp 127.0.0.1:54261->127.0.0.1:26257: write: protocol wrong type for socket
ERROR: cannot dial server.
Is the server running?
If the server is running, check --host client-side and --advertise server-side.
write tcp 127.0.0.1:54261->127.0.0.1:26257: write: protocol wrong type for socket
Failed running "sql"

**Environment:**
 - CockroachDB version [v20.1.3]
 - Server OS: [Linux/MacOS]
 - Client app [cockroach sql]

**Additional context**

Unable to load backups made by cockroach dump

**Comment by lucy-zhang** [comment]
Doesn't seem like a schema problem, I'm taking this off our board unless Bulk IO determines otherwise.

**Comment by dt** [comment]
@zsalab Thanks for the detailed bug report and thorough steps to reproduce! that's always much appreciated!

@lucy-zhang yeah, sounds right -- this isn't schema. Looks like this is either bulk-io for putting too many rows per insert, or sql (exec?) for not handling the large insert statement in the incoming SQL connection.

It looks like that dd call is bs=1k and count=8192, so sounds like 8mb rows and 100 rows per insert means an 800mb SQL statement? That seems unlikely to be something our connection buffer/lexer/parser will be okay with. That probably implies we want a knob to control rows/insert emitted in `dump`, instead of this constant: 
https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/dump.go#L775

**Comment by dt** [comment]
That knob might look something like: #52047

**Comment by dt** [comment]
looks like we stumbled across this some time ago but didn't fix it then #31676 😬

I'm going to go ahead and close this as a duplicate of #31676 but thank you again @zsalab for the great report and reproduction script!
